kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -311,29 +478,79 @@ public void initiateConnect(HostPort remote, List<FilterAndInvoker> filters) {
             pipeline.addFirst("networkLogger", new LoggingHandler("io.kroxylicious.proxy.internal.UpstreamNetworkLogger"));
         }
 
-        virtualCluster.getUpstreamSslContext().ifPresent(c -> pipeline.addFirst("ssl", c.newHandler(outboundChannel.alloc(), remote.host(), remote.port())));
-
-        connectFuture.addListener(future -> {
+        virtualCluster.getUpstreamSslContext().ifPresent(c -> {
+            final SslHandler handler = c.newHandler(outboundChannel.alloc(), remote.host(), remote.port());
+            pipeline.addFirst("ssl", handler);
+            sslHandshakeFuture.set(handler.handshakeFuture());
+            registerExceptionResponse(SSLHandshakeException.class, throwable -> {
+                Object result;
+                final Object triggerMsg = bufferedMsgs != null ? bufferedMsgs.get(0) : null;
+                if (triggerMsg instanceof final DecodedRequestFrame<?> triggerFrame) {
+                    result = buildErrorResponseFrame(triggerFrame, new UnknownServerException());
+                }
+                else {
+                    result = null;
+                }
+                return Optional.ofNullable(result);
+            });
+        });
+        // If we have an SSL future we want to wait for completion of the handshake before declaring the channel active.
+        final Future<Channel> handshakeFuture = sslHandshakeFuture.get();
+        Future<?> kafkaAvailableFuture = handshakeFuture != null ? handshakeFuture : tcpConnectFuture;
```

Comment:
I think you can avoid the atomic by using a map/orElse
```suggestion
        Future<?> kafkaAvailableFuture = virtualCluster.getUpstreamSslContext().<Future<?>>map(c -> {
            final SslHandler handler = c.newHandler(outboundChannel.alloc(), remote.host(), remote.port());
            pipeline.addFirst("ssl", handler);
            sslHandshakeFuture.set(handler.handshakeFuture());
            registerExceptionResponse(SSLHandshakeException.class, throwable -> {
                Object result;
                final Object triggerMsg = bufferedMsgs != null ? bufferedMsgs.get(0) : null;
                if (triggerMsg instanceof final DecodedRequestFrame<?> triggerFrame) {
                    result = buildErrorResponseFrame(triggerFrame, new UnknownServerException());
                }
                else {
                    result = null;
                }
                return Optional.ofNullable(result);
            });
            return handler.handshakeFuture();
        }).orElse( tcpConnectFuture);
        
        kafkaAvailableFuture.addListener(future -> {
            if (future.isSuccess()) {
                onConnection(filters);
            }
            else {
                onConnectionFailed(remote, future, inboundChannel);
            }
        });
```

Reply:
Its not quite so simple and elegant as this doesn't survive contact with the type system. `sslHandshakeFuture` is a `Future<Channel>` and `tcpConnectFuture` is a `ChannelFuture`. `ChannelFuture` extends `Future<Void>`.

We've got to adapt one to the other...

Reply:
> We've got to adapt one to the other

That's why I declared the local var as `Future<?>`.  It seem to compile for me.  I didn't think too hard.

Reply:
Hmm interesting, IntelliJ was having none of it for me.

Reply:
AHHHH
Found it, I'd missed the generics on the call to `map` and was thus having trying to have map and else trying to return different types. 

Thanks for the fix.